### ISSUE TITLE
Strange behaviour of surface rendering transparency

### ISSUE THREAD

**Issue by jni** [napari/napari#673]
## 🐛 Bug

Depending on the angle I'm looking at a surface, the surface opacity shows the surfaces behind it, or doesn't.

## To Reproduce

View a surface layer, reduce the opacity, switch to 3D mode, and rotate it. See #672 for sample code to start viewing a 3D surface.

**Comment by sofroniewn** [comment]
Agreed something seems off here - you edited your example in #672 to make it more minimal - but the original was something like this:

```python
import numpy as np
from skimage import data, measure
import napari
import scipy.ndimage as ndi


image = data.binary_blobs(128, n_dim=3)
verts, faces, normals, values = (
    measure.marching_cubes_lewiner(image.astype(float), level=0.7)
)
labels = ndi.label(image)[0]
vertex_labels = ndi.map_coordinates(labels, verts.T, order=0).astype(int)


with napari.gui_qt():

    viewer = napari.view_surface((verts, faces, values))
    surf_layer = viewer.add_surface((verts, faces, vertex_labels),
                                    colormap='gist_earth')
    viewer.dims.ndisplay = 3
```

for which I see
<img width="1200" alt="Screen Shot 2019-11-11 at 2 26 25 PM" src="https://user-images.githubusercontent.com/6531703/68626056-67692580-048f-11ea-82b2-097571024061.png">

Not quite sure what I'm expecting to see / what the desired behavior is or if this is a correct minimal example.

My comment in #683 is really about the z-ordering of objects, less about transparency, so might not be so relevant

**Comment by sofroniewn** [comment]
Note when looking into this we should also look into https://github.com/napari/napari/issues/1566 as it also raises some transparency screenshot issues that might be related

**Comment by dalthviz** [comment]
Hi, just in case, checking this with a more recent main (e8c0f37b0252) and changing the example code at https://github.com/napari/napari/issues/673#issuecomment-552645941 for the following (using the `skimage.measure.marching_cubes` function instead of `skimage.measure.marching_cubes_lewiner` which is private now):

```python
from skimage import data, measure
    import napari
    import scipy.ndimage as ndi
   
   
    image = data.binary_blobs(128, n_dim=3)
    verts, faces, normals, values = (
        measure.marching_cubes(image.astype(float), level=0.7)
    )
    labels = ndi.label(image)[0]
    vertex_labels = ndi.map_coordinates(labels, verts.T, order=0).astype(int)
   
   
    with napari.gui_qt():
   
        viewer = napari.view_surface((verts, faces, values))
        surf_layer = viewer.add_surface((verts, faces, vertex_labels),
                                        colormap='gist_earth')
        viewer.dims.ndisplay = 3

```

I got the following:

![3d](https://user-images.githubusercontent.com/16781833/211955149-ec454d93-3f4d-41ec-b753-806e24a5cb1e.gif)

**Comment by brisvag** [comment]
As far as I can tell, this is what's happening:

- the bottom layer is irrelevant (remove it and you get the same behaviour)
- the `contrast_limits_range` is set by default to `(1, 5)`
- all vertices have a value of 1 except for those few colored blobs
- thus with that colormap, everything results in black except those few colored blobs (if you change to an inverted one like gray_r, it will be ok)

This can be solved by calculating correctly the initial range like we do in images. However, you'd still get opaque blackness where data is, because we're not discarding fragments at the 0 of the colormap, but painting them black. To get them to be transparent, we need a similar thing to what we recently did in #4556 (which may require some vispy work, but nothing crazy).

There's an extra "issue" which is the camera-direction-dependent change in the shading due to the light; afaik that should not happen, but I'm not super familiar with that. Maybe @kevinyamauchi can chip in, I think he worked on that?
